src/System.IO.FileSystem/src/System/IO/PathHelpers.cs

```diff
@@ -38,5 +41,149 @@ internal static bool EndsInDirectorySeparator(string path)
+                return second.Length == 0
+                    ? string.Empty
+                    : new string(second);
+
+            if (second.Length == 0)
+                return first;
+
+            string result = CombineNoChecksInternal(first.AsReadOnlySpan(), second);
+            return result;
+        }
+
+        /// <summary>
+        /// Combines two paths. Does no validation of paths, only concatenates the paths
+        /// and places a directory separator between them if needed.
+        /// </summary>
+        internal static string CombineNoChecks(ReadOnlySpan<char> first, ReadOnlySpan<char> second)
+        {
+            if (first.Length == 0)
+                return second.Length == 0
+                    ? string.Empty
+                    : new string(second);
+
+            if (second.Length == 0)
+                return new string(first);
+
+            string result = CombineNoChecksInternal(first, second);
+            return result;
+        }
+
+        /// <summary>
+        /// Combines three paths. Does no validation of paths, only concatenates the paths
+        /// and places a directory separator between them if needed.
+        /// </summary>
+        internal static string CombineNoChecks(string first, ReadOnlySpan<char> second, ReadOnlySpan<char> third)
+        {
+            if (string.IsNullOrEmpty(first))
+                return CombineNoChecks(second, third);
+
+            if (second.Length == 0)
+                return CombineNoChecks(first, third);
+
+            if (third.Length == 0)
+                return CombineNoChecks(first, second);
+
+            return CombineNoChecksInternal(first.AsReadOnlySpan(), second, third);
+        }
+
+        /// <summary>
+        /// Combines two paths. Does no validation of paths, only concatenates the paths
+        /// and places a directory separator between them if needed.
+        /// </summary>
+        internal unsafe static string CombineNoChecks(string first, string second)
+        {
+            if (string.IsNullOrEmpty(first))
+                return string.IsNullOrEmpty(second) ? string.Empty : second;
+
+            if (string.IsNullOrEmpty(second))
+                return first;
+
+            string result = CombineNoChecksInternal(first.AsReadOnlySpan(), second.AsReadOnlySpan());
+            return result;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe static string CombineNoChecksInternal(ReadOnlySpan<char> first, ReadOnlySpan<char> second)
+        {
+            Debug.Assert(first.Length > 0 && second.Length > 0, "should have dealt with empty paths");
+
+            bool hasSeparator = first[first.Length - 1] == Path.DirectorySeparatorChar
+                || second[0] == Path.DirectorySeparatorChar;
+
+            int totalLength = first.Length + second.Length + (hasSeparator ? 0 : 1);
+            string fullPath = new string('\0', totalLength);
+            fixed (char* f = fullPath)
```

Comment:
Sight...
I am wondering whether the new `string stringCreate<TState>(int length, TState state, System.Buffers.SpanAction<char, TState> action)` API is good since you cannot really use it here.

Reply:
> since you cannot really use it here

Because you can't pass spans in generics?  We're going to have that problem in a bunch of places and is something we should try to solve more generally, e.g. a "ref struct" constraint and then an overload that accepts a delegate that takes such a constrained arg (and maybe a tuple of ref structs).

You can use the overload here, though, it just means pinning the input spans instead of the destination string, e.g.
```C#
using (char* firstPtr = &first.DangerousGetPinnableReference(), secondPtr = &second.DangerousGetPinnableReference())
{
    return string.Create(totalLength, ((IntPtr)firstPtr, first.Length, (IntPtr)secondPtr, second.Length, hasSeparator), (dst, state) =>
    {
        new Span<char>((char*)state.Item1, state.Item2).CopyTo(dst);
        if (!state.Item5) dst[state.Item2] = Path.DirectorySeparatorChar;
        new span<char>((char*)state.Item3, state.Item4).CopyTo(dst.Slice(state.Item2 + (state.Item5 ? 0 : 1));
    });
}
```
etc.  And tuple arg names could be used to avoid those ItemXs.

Reply:
I'll try using string.Create